internal/core/src/common/ArrayOffsets.cpp

```diff
@@ -0,0 +1,268 @@
+                if (valid_flags.empty() || valid_flags[i]) {
+                    array_len = array_views[i].length();
+                }
+
+                // Record the start position for this row
+                row_to_element_start[current_row_id] = element_row_ids.size();
+
+                // Add row_id for each element (elem_idx computed on access)
+                for (int32_t j = 0; j < array_len; ++j) {
+                    element_row_ids.emplace_back(current_row_id);
+                }
+
+                current_row_id++;
+            }
+        }
+    }
+
+    // Store total element count as the last entry
+    row_to_element_start[row_count] = element_row_ids.size();
+
+    AssertInfo(current_row_id == row_count,
+               "Row count mismatch: expected {}, got {}",
+               row_count,
+               current_row_id);
+
+    int64_t total_elements = element_row_ids.size();
+
+    LOG_INFO(
+        "ArrayOffsetsSealed::BuildFromSegment: struct_name='{}', "
+        "field_id={}, row_count={}, total_elements={}",
+        field_meta.get_name().get(),
+        field_meta.get_id().get(),
+        row_count,
+        total_elements);
+
+    return ArrayOffsetsSealed(std::move(element_row_ids),
+                              std::move(row_to_element_start));
+}
+
+std::pair<int64_t, int64_t>
+ArrayOffsetsGrowing::ElementIDToRowID(int64_t elem_id) const {
+    std::shared_lock lock(mutex_);
+    assert(elem_id >= 0 &&
+           elem_id < static_cast<int64_t>(element_row_ids_.size()));
+    int32_t row_id = element_row_ids_[elem_id];
+    // Compute elem_idx: elem_idx = elem_id - start_of_this_row
+    int32_t elem_idx = elem_id - row_to_element_start_[row_id];
+    return {row_id, elem_idx};
+}
+
+std::pair<int64_t, int64_t>
+ArrayOffsetsGrowing::ElementIDRangeOfRow(int64_t row_id) const {
+    std::shared_lock lock(mutex_);
+    assert(row_id >= 0 && row_id <= committed_row_count_);
+
+    if (row_id == committed_row_count_) {
+        auto total = row_to_element_start_[committed_row_count_];
+        return {total, total};
+    }
+    return {row_to_element_start_[row_id], row_to_element_start_[row_id + 1]};
+}
+
+std::pair<TargetBitmap, TargetBitmap>
+ArrayOffsetsGrowing::RowBitsetToElementBitset(
+    const TargetBitmapView& row_bitset,
+    const TargetBitmapView& valid_row_bitset) const {
+    std::shared_lock lock(mutex_);
+
+    int64_t element_count = element_row_ids_.size();
+    TargetBitmap element_bitset(element_count);
+    TargetBitmap valid_element_bitset(element_count);
+
+    // Direct access to element_row_ids_, no virtual function calls
+    for (size_t elem_id = 0; elem_id < element_row_ids_.size(); ++elem_id) {
+        auto row_id = element_row_ids_[elem_id];
+        element_bitset[elem_id] = row_bitset[row_id];
+        valid_element_bitset[elem_id] = valid_row_bitset[row_id];
+    }
+
+    return {std::move(element_bitset), std::move(valid_element_bitset)};
+}
+
```

Comment:
maybe DEBUG level

Reply:
<img width="911" height="136" alt="image" src="https://github.com/user-attachments/assets/a9551bcb-b761-49b6-88ba-2b88a910c7c8" />
You probably mean: LOG_INFO(
        "ArrayOffsetsSealed::BuildFromSegment: struct_name='{}', "
        "field_id={}, row_count={}, total_elements={}",
        field_meta.get_name().get(),
        field_meta.get_id().get(),
        row_count,
        total_elements);
        
I think it's fine? It's per segment loading and will not overwhelm